firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java

```diff
@@ -317,36 +327,208 @@ void recalculateAndSaveOverlays(Set<DocumentKey> documentKeys) {
     }
   }
 
-  /** Queries the remote documents and overlays by doing a full collection scan. */
+  /**
+   * Given a collection group, returns the next documents that follow the provided offset, along
+   * with an updated offset.
+   *
+   * <p>The first call to this method should set {@code updateMemoizedResults} to {@code true} to
+   * load remote documents and mutations into persistence. Subsequent calls to this method should be
+   * made with {@code updateMemoizedResults} set to {@code false} {@code offset} set to the offset
+   * returned by the previous call.
+   *
+   * <p>This method returns documents in order of read time from the provided offset, followed by
+   * documents with mutations in order of batch id from the offset.
+   *
+   * <p>If no documents are found, returns an empty list and an offset with the latest read time in
+   * the remote document cache.
+   *
+   * @param collectionGroup The collection group for the documents.
+   * @param offset The offset to index into.
+   * @param updateMemoizedResults Whether to load documents and mutations from persistence.
+   * @return A pair containing the next offset and a list of documents that follow the provided
+   *     offset.
+   */
+  public Pair<IndexOffset, List<Document>> getNextDocumentsAndOffsetForCollectionGroup(
```

Comment:
This would be way simpler if Document contained its BatchId. Here is a strawman proposal: https://gist.github.com/schmidt-sebastian/198aa9a0c44b928abe5fc8e5238b584e

This basically means that we can remove all IndexOffset computation from LocalDocumentsView and keep it in the IndexBackfiller.

Reply:
Implemented, thanks for starting